Introduction to Matilda
========================================

Recent development of multimodal single-cell omics tech nologies enables multiple modalities of cellular regulatory circuitry to be simultaneously profiled in individual cells (1). Data generated from these technologies create new op portunities for integrative analysis of cellular programs that are inaccessible from analysing each data modality alone and hence promise to provide a more holistic characteriza tion of cellular systems at single-cell resolution (2). A large numberofcomputational methodshavebeendevelopedfor single-cell RNA-sequencing (scRNA-seq) data to perform tasks such as data simulation (3), dimension reduction (4) and classification of cell types (5,6), and feature selection (7,8). While methods designed for scRNA-seq data analy sis can be applied to analyse RNA modality in multimodal single-cell omics data, most of them cannot take advantage of other available data modalities and therefore could not fully utilize the information embedded in such data. The lack of computational methods that can integrate across datamodalitiesisakeyissueinmultimodalsingle-cellomics data analysis and greatly hinder biological discovery from such data (9,10). 

Here we present Matilda, a neural network-based multi task learning method for integrative analysis of multimodal single-cell omics data (Figure 1A). Although previously methods developed for scRNA-seq data analysis typically address different tasks (e.g. data simulation, cell type clas sification) independently, a key observation in Matilda is that many common tasks in multimodal single-cell omics data analysis are closely related to each other. The modular ity nature of neural networks employed in Matilda makes it well-suited for integrating multiple data modalities and performing multiple tasks in a single unified framework. For example, the data simulated by the variational au toencoder (VAE) (11), a key component of Matilda, can be augmented to the original data to improve cell type classification. By leveraging such relationships, Matilda si multaneously performs data simulation, dimension reduc tion, cell type classification, and feature selection across data modalities (Figure 1A), therefore, achieving multiple key tasks in integrative analysis of multimodal single-cell omics data. 

Matilda performs data simulation, cell type classifica tion, and feature selection for single-cell multimodal omics data in a single multi-task learning framework. To evaluate the performance of Matilda on multiple tasks in multimodal single-cell omics data analysis, we applied Matilda to a collection of datasets generated from pop ular multimodal single-cell omics technologies including those profiling three modalities using TEA-seq (gene ex pression [RNA],cellsurfaceproteins[ADT],andchromatin accessibility [ATAC]) (12), and those profiling two modali ties using CITE-seq(RNAandADT)(13â€“15)andSHARE seq (RNA and ATAC) (16). While there are currently few methods specifically designed for data simulation, cell type classification, and feature selection using multiple modal ities in these datasets, various methods (e.g. Sparsim (17) for data simulation, scClassify (5) for cell type classifi cation, MAST (8) for feature selection) have been devel oped for single-cell RNA-sequencing (scRNA-seq) data and therefore can be applied using the RNA modality in these datasets. Using a range of evaluation criteria, we show that Matilda outperforms other state-of-the-art method ologies designed for various tasks using single or multi ple data modalities. Our results demonstrate the utility of Matilda as the first comprehensive method for address ing multiple key tasks in multimodal single-cell omics data analysis.


--------------



,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
