Workflow
======================================================

In this introductory tutorial, we basically go through the different steps of a complete Matilda's workflow.

.. image:: main.jpg
   :alt: The framework of Matilda
   :scale: 25%
   :align: center

1. Data Preperation  
--------------------------------------
Matilda could handle the sequencing data of RNA, ADT and ATAC. It expects raw count data for RNA and ADT modalities and the 'gene activity score' generated by Seurat from raw count data. Before modelling, you could consider remove part of the Genes with fewer than x% quantifications across cells in each of the three modalities. The 'x' is based on your criteria. Specific cell types could also be considered removing if they are extremely less or not important. 

2. Data Simulation and aggregation 
-----------------------------------------------
The first process of Matilda is the data simulation. As mentioned in the :doc:`DESIGN` and shown as the figure above, it is operated by the VAE in model. The multimodal data through encoder would then be integrated for joint learning. Then the latent space could be learned before decoder or classification. For the branch of decoder, it consists of multiple decoders each corresponds to an input data modality. Each decoder consists of one fully-connected layer to the output layer that has the same number of neurons as the features in the corresponding data modality. Then the data could be reconstructed and simulated.

3. Data augmentaion and balancing
--------------------------------------
As mentioned in :doc:`DESIGN`, recalling here. During the model training process, Matilda performs data augmentation and balancing using simulated data from the VAE component. Specifically, Matilda first ranks the cell types in the training dataset by the number of cells in each type. The cell type corresponding to the median number is used as the reference and those that have smaller numbers of cells are augmented to have the same number of cells as the median using VAE simulated single-cell multimodal data for each cell type. Cell types that have larger numbers of cells than the median number are randomly down-sampled to match the median number of cells as well. This strategy helps Matilda to mitigate imbalanced cell type distribution in the data (21) and better learn the molecular features of under-represented and rare cell types.

4. Dimension Reduction
--------------------------------------
During model training, Matilda learns to combine and reduce the feature dimensions of multimodal single-cell omics data to a latent space using its VAE component in the framework. The trained VAE of Matilda thus can be used for multimodal feature integration and dimension re duction of both the training and new data. 

5. Classification
--------------------------------------
The classification would be implemented after the the learning of latent space, the process would be operated with fully-connected neural network, involved with batch normalisation, dropout and so on.


6. Feature selection 
--------------------------------------
Finally, as mentioned in :doc:`DESIGN`, the neural network trained for cell type classification in Matilda can be used for multimodal feature selection using methods such as integrated gradient (IG) descent (22) and saliency procedures (23), and thus can lead to the selection of cell-type-specific features across all available modalities in the datasets. Within the two feature selection methods implemented in Matilda, IG appears to perform slightly better than saliency and is hence the recommended approach in Matilda for feature selection from multimodal single-cell omics data []. 

Figure 5A, B visualize top-ranked features selected by Matilda using IG for CD14 monocytes and Naive B cells, respectively, in each data modality in the TEA-seq dataset. The RNA and ADT expression levels and the ATAC activity of selected genes across all cell types in the dataset are shown in Figure 5C, D. As expected, these analyses reveal that features selected by Matilda for each data modality show expression specificity towards their respective cell types, demonstrating their potential usage for characterizing cell identity and their underlying molecular programs.
To evaluate the top features selected by Matilda across multiple data modalities and those selected from RNA modality using popular methods such as t-test and limma (7), and those specifically designed for scRNA-seq (e.g. MAST(8), ROC), and recently proposed deep learning feature selection methods, including PROPOSE (35) and scCapsNet (41), we compared their utility in classifying each cell type in each dataset. We found that cell-type-specific features selected by Matilda from multiple data modalities on average resulted in more accurate discrimination of their respective cell types as shown by the scatter plot and the overall rankings of methods in each dataset (Figure 5Eand Supplementary Figure S8). Within the two feature selection methods implemented in Matilda, IG appears to perform slightly better than saliency and is hence the recommended approach in Matilda for feature selection from multimodal single-cell omics data. Together, these results demonstrate Matilda as a useful approach for feature selection from multiple data modalities for cell type characterization and other downstream analyses.


7. Training continue
-----------------------------


Reference
--------------
[1]

[2]

[3]






